File: scheduler/scheduler/core/static/js/supervisor/events.js
import { csrfToken } from './config.js';
import { formatMin, timeToMin } from './utils.js';
import { loadSectionsIntoDropdown } from './roles.js';

export function populateTimeDropdowns() {
    const startSelect = document.getElementById('eventStartTime');
    const endSelect = document.getElementById('eventEndTime');
    if (!startSelect || !endSelect) return;

    for (let h = 7; h <= 22; h++) {
        for (let m = 0; m < 60; m += 15) {
            const totalMin = h * 60 + m;
            const label = formatMin(totalMin);
            startSelect.appendChild(new Option(label, totalMin));
            endSelect.appendChild(new Option(label, totalMin));
        }
    }
    endSelect.selectedIndex = 1;
}

export async function addFixedEvent() {
    const name = document.getElementById('eventNameInput').value.trim();
    const location = document.getElementById('eventLocationInput').value.trim();
    const roleId = document.getElementById('eventRoleInput').value;
    const section = document.getElementById('eventSectionInput').value || null;
    const startMin = timeToMin(document.getElementById('eventStartTime').value);
    const endMin = timeToMin(document.getElementById('eventEndTime').value);
    const checkedDays = [...document.querySelectorAll('.day-check input:checked')].map(cb => parseInt(cb.value));

    if (!name) return alert('Please enter an event name.');
    if (checkedDays.length === 0) return alert('Please select at least one day.');
    if (endMin <= startMin) return alert('End time must be after start time.');

    const dayNames = ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat'];
    const dayLabel = checkedDays.map(d => dayNames[d]).join(', ');

    try {
        const response = await fetch(`/api/team/${window.TEAM_ID}/obstructions/create/`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json', 'X-CSRFToken': csrfToken },
            body: JSON.stringify({ name, location, role_id: roleId, section, start_min: startMin, end_min: endMin, days: checkedDays })
        });
        const data = await response.json();
        if (response.ok) {
            const tag = document.createElement('span');
            tag.className = 'role-tag';
            tag.dataset.obstructionId = data.obstruction_id;
            tag.innerHTML = `<strong>${name}</strong>&nbsp;<em>${location}</em>&nbsp;${dayLabel} &bull; ${formatMin(startMin)} - ${formatMin(endMin)}<i class="fa-solid fa-xmark" onclick="deleteObstruction(this)"></i>`;
            document.getElementById('fixed-events-list').appendChild(tag);
            document.getElementById('eventNameInput').value = '';
            document.querySelectorAll('.day-check input').forEach(cb => (cb.checked = false));
        }
    } catch (error) {
        console.error(error);
    }
}

export async function deleteObstruction(icon) {
    const tag = icon.closest('.role-tag');
    const obstructionId = tag.dataset.obstructionId;
    if (!confirm('Remove this obstruction?')) return;

    try {
        const response = await fetch(`/api/team/${window.TEAM_ID}/obstructions/${obstructionId}/delete/`, {
            method: 'DELETE',
            headers: { 'X-CSRFToken': csrfToken }
        });
        if (response.ok) tag.remove();
    } catch (error) {
        console.error(error);
    }
}

export async function onEventRoleChange(select) {
    const roleId = select.value;
    const sectionSelect = document.getElementById('eventSectionInput');
    if (!roleId) {
        sectionSelect.style.display = 'none';
        sectionSelect.innerHTML = '<option value="">All sections</option>';
        return;
    }
    await loadSectionsIntoDropdown(roleId, sectionSelect);
    sectionSelect.options[0].textContent = 'All sections';
}

File: scheduler/scheduler/core/views/events.py
'''
views/events.py
API endpoints for fixed events (obstructions) and team member removal.
'''

import json
from datetime import time

from django.shortcuts import get_object_or_404
from django.http import JsonResponse, HttpResponseForbidden
from django.contrib.auth.decorators import login_required
from django.contrib.auth.models import User
from django.views.decorators.http import require_http_methods
from django.views.decorators.csrf import csrf_exempt

from ..models import (
    FixedObstruction,
    ObstructionDay,
    Role,
    RoleSection,
    Team,
    TeamEvent,
    TeamRoleAssignment,
)


@login_required
@require_http_methods(["POST"])
def add_event(request, team_id):
    team = get_object_or_404(Team, id=team_id)
    if team.owner != request.user:
        return HttpResponseForbidden("Unauthorized")

    try:
        data = json.loads(request.body)
        event = TeamEvent.objects.create(
            team=team,
            name=data.get("name"),
            day=data.get("day"),
            start_time=data.get("start"),
            end_time=data.get("end"),
        )
        return JsonResponse({"ok": True, "event": {"id": event.id, "name": event.name, "day": event.day}})
    except Exception as e:
        return JsonResponse({"ok": False, "error": str(e)}, status=400)


@csrf_exempt
@require_http_methods(["POST"])
def create_obstruction(request, team_id):
    '''
    Create a fixed obstruction (e.g. a class or commitment) that blocks scheduling for a role.
    '''
    data = json.loads(request.body)
    team = get_object_or_404(Team, id=team_id)
    role = get_object_or_404(Role, id=data["role_id"]) if data.get("role_id") else None

    start_min  = data["start_min"]
    end_min    = data["end_min"]
    start_time = time(start_min // 60, start_min % 60)
    end_time   = time(end_min // 60, end_min % 60)

    section_id  = data.get('section', None)
    section_obj = get_object_or_404(RoleSection, id=section_id) if section_id else None
    section_name = section_obj.name if section_obj else None

    obstruction = FixedObstruction.objects.create(
        team=team,
        location=data["location"],
        role=role,
        name=data["name"],
        start_time=start_time,
        end_time=end_time,
        section=section_name,
    )

    day_names = ["sun", "mon", "tue", "wed", "thu", "fri", "sat"]
    for day_index in data["days"]:
        ObstructionDay.objects.create(obstruction=obstruction, day=day_names[day_index])

    return JsonResponse({"status": "ok", "obstruction_id": obstruction.id})


@csrf_exempt
@require_http_methods(["DELETE"])
def delete_obstruction(request, team_id, obstruction_id):
    obstruction = get_object_or_404(FixedObstruction, team=team_id, id=obstruction_id)
    obstruction.delete()
    return JsonResponse({"status": "ok"})


@require_http_methods(["POST"])
def remove_member_from_team(request, team_id):
    '''
    Remove a member from the team and delete their role assignment.
    '''
    try:
        data    = json.loads(request.body)
        user_id = data.get('user_id')

        team = Team.objects.get(id=team_id)
        team.members.remove(user_id)
        TeamRoleAssignment.objects.filter(team=team, user_id=user_id).delete()

        return JsonResponse({'status': 'success'})
    except Exception as e:
        return JsonResponse({'status': 'error', 'message': str(e)}, status=400)